cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/io/cdap/cdap/etl/lineage/LineageOperationsProcessor.java

```diff
@@ -121,8 +122,12 @@ private Map<String, Operation> computeProcessedOperations() {
         switch (fieldOperation.getType()) {
           case READ:
             FieldReadOperation read = (FieldReadOperation) fieldOperation;
+            Map<String, String> sourceProperties = read.getSource().getProperties();
+            sourceProperties.put("stageName", stageName);
```

Comment:
It's not good to directly modify read.getSource().getProperties() like this. Getters should almost always return unmodifiable collections, otherwise anything that calls the getter can modify internal state.

However, I do see the EndPoint is wrapping the properties in an unmodifiableMap, so I don't really understand how this isn't throwing an exception.

Reply:
Creating a new hashmap now